003-graphql.md

```diff
@@ -0,0 +1,36 @@
+# GraphQL
+
+We should use GraphQL for all data-fetching in the future.
+
+**Discussion:** https://github.com/DoSomething/rfcs/pull/3
+
+## Problem(s)
+
+There were a couple of problems we looked to GraphQL to help solve:
+
+ - We reimplemented cross-application includes (such as `?include=user` on a list of posts in Rogue) in multiple apps. This caused different backend services to have to act as a "gateway" depending on where a request started.
+ - We found that we would eventually implement the same batching & caching logic in each application (for example, fetching campaigns [in Rogue](https://git.io/fjjGZ) or users [in our now-defunct competitions app](https://github.com/DoSomething/gladiator/blob/534ca9834a5369ededaed56deff33404bbc94e99/app/Repositories/DatabaseUserRepository.php#L62-L163)). In each case, we'd have to reconsider the right order-of-operations and maximum batch size for good performance.
+ - Because each application maintained it's own caches, it was hard to maintain a consistent cache policy. (For example, one application might cache users for an hour while another didn't cache them at all).
+ - Finally, we had to do a lot of busy-work to [normalize our Redux state](https://redux.js.org/recipes/structuring-reducers/normalizing-state-shape) in Phoenix, which slowed down development of new features. (See [DoSomething/phoenix-next#767](https://github.com/DoSomething/phoenix-next/pull/767) and [DoSomething/phoenix-next#1123](https://github.com/DoSomething/phoenix-next/pull/1123)).
+
+## Proposal
+
+We built our GraphQL gateway (see [graphql.dosomething.org](https://graphql.dosomething.org/)) in November 2017. We have gradually added more resources (such as Contentful spaces and embed metadata) and applications (such as Gambit) over time.
+
+This gives us a single [universal schema](https://principledgraphql.com/integrity#1-one-graph) (so developers don't need to worry about which application holds which data). It lets us optimally [batch requests to backend services](https://www.apollographql.com/docs/graphql-tools/connectors/#dataloader-and-caching) and acts as a centralized "wrapper" around third-party APIs like [Contentful](https://www.contentful.com/).
```

Comment:
> This gives us a single [universal schema](https://principledgraphql.com/integrity#1-one-graph) (so developers don't need to worry about which application holds which data).

Perhaps this (the issue of multiple API's needing to be stitched together awkwardly) can be stressed more as a pain point in the `Problems` section? Maybe added to the first point there? (I always saw this as one of the main reasons for our adoption, but didn't feel the argument strongly in the problems section, though that could just be a personal misunderstanding.)

Reply:
Yeah, that was definitely missing! Added as the first bullet point in b01c6f6.

Reply:
Another benefit that's perhaps assumed by having a single universal schema, but maybe worth spelling out here is that there will only be a single source of authentication for the DS API. Gambit, for example, is an app that stores authentication credentials for GraphQL, Northstar (via [Northstar JS](5d83c1a93d6e6a009c355da2) -- I've long wanted to replace this package for GraphQL), and Rogue -- only because we haven't swapped our GET/POST requests for GraphQL equivalents.

Also - back in the old days when I was working on the Gambit+GraphQL integration, I was blocked by using GraphQL for users because of authenticating as an admin machine. Is that still the case? Ideally Gambit Admin would pull message history from GraphQL as well (but I never added Gambit messages into GraphQL because of the authentication issues)

Reply:
> Also - back in the old days when I was working on the Gambit+GraphQL integration, I was blocked by using GraphQL for users because of authenticating as an admin machine. Is that still the case? Ideally Gambit Admin would pull message history from GraphQL as well (but I never added Gambit messages into GraphQL because of the authentication issues)

Yeah, if I remember the issue correctly, this is still the case. GraphQL "supports" Northstar's OAuth tokens for Northstar & Rogue simply because it forwards them along on the backend requests (and all the authorization logic happens within the relevant service).

Since Gambit doesn't support authorizing requests via OAuth tokens, we'd have to add support to GraphQL to parse those tokens & use that to determine whether or not to use the static Gambit API key for the backend request.